Optimized SmartLabel type

YohDeadfall opened the pull request:
This PR turn the `SmartLabel` type into a slim wrapper around a backing array, so the type can fit into a single register. Additionally I changed `Labels` property type from `IEnumerable<string>` to `ReadOnlySpan<string>` which allows no allocating iteration over elements. The same happens when `foreach` loop is used over an instance itself because `GetEnumerator` returns `ReadOnlySpan<string>.Enumerator` while `IEnumerable<string` interface is implemented explicitly.

The only controversial thing I see here is the type of `Labels` property or maybe its existence. A better option, I guess, would be a method named `AsSpan()`.

YohDeadfall commented on WalletWasabi/Blockchain/Analysis/Clustering/SmartLabel.cs:
```diff
@@ -1,153 +1,168 @@
 namespace WalletWasabi.Blockchain.Analysis.Clustering;
 
-public class SmartLabel : IEquatable<SmartLabel>, IEquatable<string>, IEnumerable<string>, IComparable<SmartLabel>
+public readonly struct SmartLabel : IEquatable<SmartLabel>, IComparable<SmartLabel>, IEnumerable<string>
```
True, but that is out of this PR scope.

wieslawsoltes replied:
I would rename it to `LabelsArray`

YohDeadfall replied:
Can it be done in a separate PR, or would you like to see it here?

wieslawsoltes replied:
Can be separate PR

YohDeadfall replied:
Then review this PR, please, so I can rename it separately.

wieslawsoltes commented on WalletWasabi/Blockchain/Analysis/Clustering/SmartLabel.cs:
```diff
@@ -1,153 +1,168 @@
+		Labels.GetEnumerator();
 
-	public static bool operator ==(string? x, SmartLabel? y) => AreEquivalent(x, y);
+	IEnumerator<string> IEnumerable<string>.GetEnumerator() => GetEnumeratorAllocating();
```
I don't see any unit tests for that part.

YohDeadfall replied:
That part wasn't

YohDeadfall replied:
That part wasn't changed, so added no tests on it, but you're right.

YohDeadfall replied:
@wieslawsoltes, I fixed it, so the pull request is waiting for another review.

YohDeadfall commented:
Ugh, I expected that bindings to `ref` types cannot be compiled in Avalonia, but it does that and does it incorrectly. Therefore, I removed the property and added a method instead of it to avoid such issues in the future.

wieslawsoltes reviewed (changes requested):
Sorting labels is now case-sensitive:

master
![image](https://user-images.githubusercontent.com/2297442/233069546-305d7466-c81c-4dc9-b0c6-76eba60618e9.png)
PR
![image](https://user-images.githubusercontent.com/2297442/233069887-40db2605-1f22-4167-8bd4-a73f1c78cc9e.png)

YohDeadfall commented:
What makes me unhappy here is an implicit conversion to `string` and difference between `Equals` and `CompareTo`.

YohDeadfall commented on WalletWasabi.Tests/UnitTests/SmartLabelTests.cs:
```diff
@@ -125,6 +163,10 @@ public void CaseSensitivityTests()
 		stringLabelToCheck = "buZ, BaR, fOo";
+		Assert.False(smartLabel.Equals(smartLabelToCheck));
+		Assert.False(smartLabel.Equals(stringLabelToCheck));
+		Assert.True(0 == smartLabel.CompareTo(smartLabelToCheck));
```
@wieslawsoltes, the default comparison is handled by tests here now.

wieslawsoltes replied:
Great

wieslawsoltes commented:
> What makes me unhappy here is an implicit conversion to `string` and difference between `Equals` and `CompareTo`.

Don't see a way to avoid it

wieslawsoltes commented:
Tested LGTM

YohDeadfall commented on WalletWasabi/Blockchain/Analysis/Clustering/SmartLabel.cs:
```diff
@@ -1,153 +1,170 @@
 
-	public bool Equals(string? other, StringComparison comparison) => AreEquivalent(other, this, comparison);
+	public bool Equals(string? other) =>
+		ToString().Equals(other);
 
-	public bool Equals(SmartLabel? other, IEqualityComparer<string> comparer) => AreEquivalent(this, other, comparer);
+	public bool Equals(string? other, StringComparison comparison) =>
+		ToString().Equals(other, comparison);
```
It's the rule of thumb to implement other methods accepting a comparer if there's one already. At the same time I understand what bothers you, that type has a lot of methods which can be removed actually, but that's not an optimization then. I mean that a thing I would like to do in a separate PR with fixing a leak of UI into that struct (comparison and equality behave differently). Had a chat with Wieslav on that already.

YohDeadfall replied:
Ah, those. Yeah, would like to drop comparisons with strings completely. Is this PR sounds okay to do that?

YohDeadfall replied:
Then will do in a follow up.

YohDeadfall commented on WalletWasabi/Blockchain/Analysis/Clustering/SmartLabel.cs:
```diff
@@ -1,153 +1,170 @@
 	}
 
-	public static SmartLabel Empty { get; } = new SmartLabel();
+	public static SmartLabel Empty { get; } = new();
```
It's a property, not a field (:

YohDeadfall commented on WalletWasabi/Blockchain/Analysis/Clustering/SmartLabel.cs:
```diff
@@ -1,153 +1,170 @@
+		var thisLabels = AsSpan();
+		var otherLabels = other.AsSpan();
+		
+		return CompareToSlow(
```
Kinda agree, but that's an approach used by Microsoft and those who follow them. If you have strong feelings about making it prefixed instead, I will do.

YohDeadfall replied:
That exact method doesn't exist, I just proposed it. Consider examples, look at [`dotnet/runtime`](https://github.com/search?q=repo%3Adotnet%2Fruntime+Slow+path%3A%2F%5Esrc%5C%2F%2F+language%3AC%23+path%3A%2F%5Esrc%5C%2Flibraries%5C%2FSystem.Private.CoreLib%5C%2Fsrc%5C%2FSystem%5C%2F%2F&type=code).

YohDeadfall replied:
Should be, the link seems fine. Okay, here's [the first one](https://github.com/dotnet/runtime/blob/9ea909abbacfe0d29cfd6358a15858316686b5fd/src/libraries/System.Private.CoreLib/src/System/IO/BufferedStream.cs#L740):

```csharp
        public override int ReadByte()
        {
            return _readPos != _readLen ?
                _buffer![_readPos++] :
                ReadByteSlow();
        }

        private int ReadByteSlow()
```

YohDeadfall replied:
There's an idea behind it - IntelliSense or any other language server provides you a slow version of the method by typing the current method name. In that case one just selects that version via arrows or by adding `S` and finishing auto completion.

YohDeadfall commented on WalletWasabi/Blockchain/Analysis/Clustering/SmartLabel.cs:
```diff
@@ -1,153 +1,170 @@
 
-	public static SmartLabel Empty { get; } = new SmartLabel();
+	public static SmartLabel Empty { get; } = new();
 	public static char[] Separators { get; } = new[] { ',', ':' };
```
Nope, it's a getter-only property.

YohDeadfall commented on WalletWasabi/Blockchain/Analysis/Clustering/SmartLabel.cs:
```diff
@@ -1,153 +1,170 @@
-	}
-
-	public static SmartLabel Merge(params SmartLabel[] labels) => Merge(labels as IEnumerable<SmartLabel>);
+		const string Separator = ", ";
```
Difference between both of them is that method-level constants doesn't produce any metadata during compilation. It doesn't matter if one doesn't care about size of the resulting assembly.

YohDeadfall replied:
One more reason here is avoid name space pollution, so things if it's not required in other places, then no need to expose it.

YohDeadfall commented on WalletWasabi.Tests/UnitTests/SmartLabelTests.cs:
```diff
@@ -12,54 +12,88 @@ public void LabelParsingTests()
 		label = new SmartLabel("");
 		Assert.Equal("", label);
+		Assert.Equal(Array.Empty<string>(), label);
+		Assert.Equal(Array.Empty<string>(), label.AsSpan().ToArray());
```
It's intentional because an enumerator produced through `IEnumerable<T>.GetEnumerator()` and `ReadOnlySpan<T>.GetEnumerator()` are completely different. The last one implements no interfaces and cannot be allocated on the heap. Therefore, these two lines test different code paths.

YohDeadfall commented on WalletWasabi.Fluent/ViewModels/Wallets/Labels/SuggestionLabelsViewModel.cs:
```diff
@@ -48,7 +48,7 @@ public void UpdateLabels()
 		// Make recent and receive labels count more for the current wallet
 		var multiplier = 100;
-		foreach (var label in KeyManager.GetReceiveLabels().Reverse().SelectMany(x => x.Labels))
+		foreach (var label in KeyManager.GetReceiveLabels().Reverse().SelectMany(x => x))
```
Not exactly, it converts a `SmartLabel` into `IEnumerable<string>`. There's a better name for the type, but it will come with a following cleanup.

YohDeadfall commented on WalletWasabi/Blockchain/Analysis/Clustering/SmartLabel.cs:
```diff
@@ -1,153 +1,172 @@
 
-	public static bool operator !=(SmartLabel? x, string? y) => !(x == y);
+	public static SmartLabel Merge(IEnumerable<SmartLabel> labels) =>
+		new(labels?.SelectMany(x => x));
```
The original code assumes that there can be a null reference:

https://github.com/zkSNACKs/WalletWasabi/blob/eb062deac486200080855bb70dcfd12cc90c23dc/WalletWasabi/Blockchain/Analysis/Clustering/SmartLabel.cs#L38-L40

YohDeadfall replied:
Let's cleanup that later too as anything else was found here?

YohDeadfall commented on WalletWasabi/Blockchain/Analysis/Clustering/SmartLabel.cs:
```diff
@@ -1,153 +1,172 @@
 	}
 
-	public SmartLabel(IEnumerable<string> labels)
+	public SmartLabel(IEnumerable<string>? labels)
```
The original code assumes that even if the argument isn't nullable, it's possible to get it:

https://github.com/zkSNACKs/WalletWasabi/blob/eb062deac486200080855bb70dcfd12cc90c23dc/WalletWasabi/Blockchain/Analysis/Clustering/SmartLabel.cs#L13-L15

Therefore, it's correct to allow a null reference to be passed. Otherwise, there should be an exception since `null` isn't allowed in case of removal of the nullability marker.
